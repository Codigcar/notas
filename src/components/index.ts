

export * from './CRenderItem';
export * from './CRenderList';
export * from './CSearchBar';
export * from './CTextCenter';