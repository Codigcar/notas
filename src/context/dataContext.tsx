import React from 'react';
import { createContext } from 'react';
import { IPublication } from '../interfaces/publish.interface';

interface ContextProps {
    publications?: IPublication[];

    getNextPublication: () => Promise<{hasError: boolean, message: string, data: IPublication[]}>
    resetData: () => Promise<void>;
    initData: () => Promise<void>;
    setStorage: (key: string, value: string) => Promise<void>;
    getStorage: (key: string) => Promise<string | null>;
    removeStorage: (key: string) => Promise<void>;

}

export const DataContext = createContext({} as ContextProps);
