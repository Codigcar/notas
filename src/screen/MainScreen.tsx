import React, {useContext, useState, useEffect} from 'react';
import {View, StyleSheet} from 'react-native';

import {CTextCenter, CRenderList, CSearchBar} from '../components/index';
import {DataContext} from '../context/index';

const MainScreen = () => {
  const [search, setSearch] = useState<string>('');
  const [filteredDataSource, setFilteredDataSource] = useState<any>('');
  const [isLoadingPullRefresh, setIsLoadingPullRefresh] = useState(false);

  const {publications, getNextPublication, resetData, initData} = useContext(DataContext);

  const loadNextPublication = () => {
    getNextPublication();
  };

  const pullRefresh = async () => {
    await resetData();
    await initData();
    setSearch('');
  };

  // View

  if (publications === undefined) {
    return <CTextCenter title="Cargando..." />;
  }

  return (
    <View style={styles.container}>
      <CSearchBar
        dataList={publications}
        search={search}
        setSearch={setSearch}
        setFilteredDataSource={setFilteredDataSource}
      />

      <View style={styles.dataList}>
        {search.trim() === '' ? (
          <CRenderList
            data={publications}
            loadMore={loadNextPublication}
            timeEndReached={0.2}
            isLoadingPullRefresh={isLoadingPullRefresh}
            pullRefresh={pullRefresh}
          />
        ) : (
          <CRenderList
            data={filteredDataSource}
            isLoadingPullRefresh={isLoadingPullRefresh}
            pullRefresh={pullRefresh}
          />
        )}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  dataList: {
    flex: 1,
  },
});

export default MainScreen;
