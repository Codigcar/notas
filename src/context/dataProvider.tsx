import React, {useEffect, useReducer} from 'react';

import AsyncStorage from '@react-native-async-storage/async-storage';
import {IPublication} from '../interfaces/publish.interface';

import {dataReducer} from './dataReducer';
import {fetchPublications} from '../api/fetchPublications';
import {DataContext} from './dataContext';

export interface IDataState {
  last_user: number,
  publications: IPublication[];
}

export const DATA_INITIAL_STATE: IDataState = {
  last_user: 1,
  publications: [],
};

// @ts-ignore
export const DataProvider: FC<any> = ({children}) => {
  const [state, dispatch] = useReducer(dataReducer,DATA_INITIAL_STATE);
  
  const key = "[state]";

  const setStorage = async(key: string, value: string): Promise<void> => {
    await AsyncStorage.setItem(key, value);
  }

  const getStorage = async(key: string): Promise<string | null> => {
    const data = await AsyncStorage.getItem(key);
    if(!data) {
      return null;
    }
    return data;
  }

  const removeStorage = async(key: string): Promise<void> => {
    await AsyncStorage.removeItem(key);
  }

  useEffect(() => {
    initData()
    return () => {}
  }, [])

  useEffect(() => {
    persistData()
    return () => {}

  }, [state]);

  

  const initData = async() => {
    getFetchDataInit().then((data: IDataState) => {
      dispatch({type:'[Publication] - init', payload: {
            last_user: data.last_user,
            publications: data.publications
      }})
    })
    .catch((err:string) => {
      console.error({err});
    })
  }

  const getFetchDataInit = async() => {
    const data = await getStorage(key);
    const dataParse = data ? JSON.parse(data) : DATA_INITIAL_STATE;
    const { last_user, publications } = dataParse;

    if(publications.length > 0) {
      return {last_user, publications};
    } 

    const getPublications = await fetchPublications(1);
    return {last_user, publications: getPublications};
  }

  
  const persistData = async() => {
    await setStorage(key, JSON.stringify(state));
  }
  
  const resetData = async() => {
    await dispatch({type:'[Publication] - init', payload: DATA_INITIAL_STATE})
  }

  const getNextPublication = async(): Promise<{hasError: boolean; message: string; data: IPublication[]}> => {
    if(state.last_user > 4) {
      return {
        hasError: false,
        message: '',
        data: [],
      };
    }
    const last_user = state.last_user + 1;
    const getPublications = await fetchPublications(last_user);

    dispatch({type:'[Publication] - AddPublications', payload: {
      last_user: last_user,
      publications: getPublications
    }})

    return {
      hasError: false,
      message: '',
      data: getPublications,
    };
  };

  return (
    <DataContext.Provider
      value={{
        ...state,

        // Methods
        getNextPublication,
        resetData,
        initData,
        getStorage,
        setStorage,
        removeStorage
      }}>
      {children}
    </DataContext.Provider>
  );
};
