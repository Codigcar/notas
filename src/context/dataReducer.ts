import { IDataState, DATA_INITIAL_STATE } from './dataProvider';
import { IData } from '../interfaces/publish.interface';

type DataActionType =
  | { type: '[Publication] - init'; payload: IData }
  | { type: '[Publication] - AddPublications'; payload: IData }
  | { type: '[Publication] - reset'; }

export const dataReducer = (state: IDataState = DATA_INITIAL_STATE, action: DataActionType): IDataState => {
    switch (action.type) {

      case '[Publication] - init':
        return {
          ...state,
          last_user: action.payload.last_user,
          publications: action.payload.publications
        };

      case '[Publication] - AddPublications':
        return {
          ...state,
          last_user: action.payload.last_user,
          publications: state.publications.concat(action.payload.publications)
        };
      
      default:
        return state;
    }
  };