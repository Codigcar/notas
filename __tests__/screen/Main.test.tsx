import React from 'react';
import {DataContext} from '../../src/context';
import MainScreen from '../../src/screen/MainScreen';

const {render} = require('@testing-library/react-native');
describe('testing en <CRenderItem />', () => {
  it('debe hacer match con el snapshot', () => {
    const component = render(<MainScreen />);
    expect(component).toMatchSnapshot();
  });

  it('debe mostrar el loading inicial', () => {
    const {getByText} = render(<MainScreen />);
    const textLoading = 'Cargando...';
    expect(getByText(textLoading)).toBeTruthy();
  });

  it('debe renderizar la lista de publicaciones obtenidas', () => {
    const publicationsList = [
      {
        userId: 1,
        id: 1,
        title: 'title1',
        body: 'body1',
      },
      {
        userId: 1,
        id: 2,
        title: 'title2',
        body: 'body2',
      },
      {
        userId: 1,
        id: 3,
        title: 'title3',
        body: 'body3',
      },
    ];
    const getNextPublication = jest.fn();
    const resetData = jest.fn();
    const initData = jest.fn();
    const getStorage = jest.fn();
    const setStorage = jest.fn();
    const removeStorage = jest.fn();

    beforeEach(() => jest.clearAllMocks());

    const {getByText} = render(
      <DataContext.Provider
        value={{
          publications: publicationsList,
          getNextPublication,
          resetData,
          initData,
          getStorage,
          setStorage,
          removeStorage,
        }}>
        <MainScreen />
      </DataContext.Provider>,
    );
    expect(getByText(publicationsList[0].title)).toBeTruthy();
  });
});
