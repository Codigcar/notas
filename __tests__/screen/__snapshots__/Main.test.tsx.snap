// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing en <CRenderItem /> debe hacer match con el snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <View>
    <Text>
      Cargando...
    </Text>
  </View>
</View>
`;
